processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -1496,48 +1488,82 @@ public boolean putToKeyBuffer(RowBasedKey key, int idx)
         return true;
       }
 
-      @Override
-      public void getFromByteBuffer(ByteBuffer buffer, int initialOffset, int dimValIdx, Comparable[] dimValues)
+      /**
+       * Adds s to the dictionary. If the dictionary's size limit would be exceeded by adding this key, then
+       * this returns -1.
+       *
+       * @param s a string
+       *
+       * @return id for this string, or -1
+       */
+      private int addToDictionary(final String s)
       {
-        dimValues[dimValIdx] = dictionary.get(buffer.getInt(initialOffset + keyBufferPosition));
-      }
+        int idx = reverseDictionary.getInt(s);
+        if (idx == UNKNOWN_DICTIONARY_KEY) {
+          final long additionalEstimatedSize = estimateStringKeySize(s);
+          if (currentEstimatedSize + additionalEstimatedSize > maxDictionarySize) {
+            return -1;
+          }
 
-      @Override
-      public int compare(ByteBuffer lhsBuffer, ByteBuffer rhsBuffer, int lhsPosition, int rhsPosition)
-      {
-        return Ints.compare(
-            sortableIds[lhsBuffer.getInt(lhsPosition + keyBufferPosition)],
-            sortableIds[rhsBuffer.getInt(rhsPosition + keyBufferPosition)]
-        );
+          idx = dictionary.size();
+          reverseDictionary.put(s, idx);
+          dictionary.add(s);
+          currentEstimatedSize += additionalEstimatedSize;
```

Comment:
Shouldn't `sortableIds` also be updated?

Reply:
sortableIds is mutable for DynamicDictionaryStringRowBasedKeySerdeHelper and immutable for StaticDictionaryStringRowBasedKeySerdeHelper. I added a comment.

Reply:
Isn't `addToDictionary()` called with dynamic dictionary?

Reply:
Yes. What do you mean by this question?

Reply:
This method is used to add a string to a dictionary, but it doesn't update `rankOfDictionaryIds`

Reply:
Ah, I got your point. I updated the comment.

Reply:
I still don't understand why `rankOfDictionaryIds` is not updated here. It doesn't need to be updated?

Reply:
Yes, it is created in bufferComparator() with dynamic dictionary and not changed dynamically. See https://github.com/druid-io/druid/pull/4704/files#diff-8da1c4dd87d76bcce224a6be6bf74ed8R969.